File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
// @Author: KING-SEN
// Programming Language Used: C++

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> dict(256, -1);
        int maxLen = 0, start = -1;
        for (int i = 0; i < s.length(); i++) {
            if (dict[s[i]] > start)
                start = dict[s[i]];
            dict[s[i]] = i;
            maxLen = max(maxLen, i - start);
        }
        return maxLen;
    }
};

// Time Complexity: O(N), Where N is length of the string 
// Space Complexity: O(min(N, M)), where N is the length of the string and M is the size of the character set for HashMap
// Space Complexity: O(m), where m is the size of the charset.

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic = [-1] * 256
        maxLen, start = 0, -1
        for i, c in enumerate(s):
            if dic[ord(c)] > start:
                start = dic[ord(c)]
            dic[ord(c)] = i
            maxLen = max(maxLen, i - start)
        return maxLen